rust/kernel/chrdev.rs

```diff
@@ -20,10 +20,66 @@ use crate::error::{Error, KernelResult};
 use crate::file_operations;
 use crate::types::CStr;
 
+struct SafeCdev(*mut bindings::cdev);
+
+impl SafeCdev {
+    fn alloc() -> KernelResult<Self> {
+        // SAFETY: call an unsafe function
+        let cdev = unsafe { bindings::cdev_alloc() };
+        if cdev.is_null() {
+            return Err(Error::ENOMEM);
+        }
+        Ok(Self(cdev))
+    }
+
+    fn init(&mut self, fops: *const bindings::file_operations) {
+        // SAFETY: call an unsafe function
+        unsafe {
+            bindings::cdev_init(self.0, fops);
+        }
+    }
+
+    fn add(&mut self, dev: bindings::dev_t, pos: c_types::c_uint) -> KernelResult<()> {
+        // SAFETY: call an unsafe function
+        let rc = unsafe { bindings::cdev_add(self.0, dev, pos) };
+        if rc != 0 {
+            return Err(Error::from_kernel_errno(rc));
+        }
+        Ok(())
+    }
+
+    fn set_owner(&mut self, module: &crate::ThisModule) {
+        // SAFETY: dereference of a raw pointer
```

Comment:
You're looking for a description of a proof-of-safety, which runs inside a human brain? :) Could you point me to an instance of this in the existing codebase? I often find it easier to work from an existing example.

No concern that new commits will break existing proofs-of-safety?

Reply:
I inherited the `SAFETY` comment practice from the Rust's standard library; and most of the code we have already follows the convention.

New commits that fail to update the comment are a bug. That is, any commit updating any `unsafe` block *must* re-check the `SAFETY` comment that goes with it. Same for `unsafe` functions and their `Safety` section. That is why having `unsafe` blocks as small as possible and as close as possible to the operation helps avoiding "comment bugs" :)

I know the C side of the kernel does not follow such strict comment guidelines, but I believe this is a very good thing. And yes, many of those comments will be trivial, but the point is not just to document what is non-trivial, but to also know there is nothing non-trivial going on.

Reply:
Yes, I agree 100% with the need for rigour here. One of the big justifications for this effort is memory safety issues, so we better get this right, and keep it right. I would like to learn what's required here - I'll see if I can make a rigorous SAFETY argument in the next push.

Reply:
As to how to write the soundness "proof", a good way is to check the documentation of what you are calling to see why it is `unsafe` (or the language reference, for other operations). Then, think why those preconditions hold, and write it down.

So, for instance, here, the comment should explain why `self.0` is guaranteed in all cases to be a valid (dereferenceable) pointer. Thus you can ask yourself the question "Is there any way someone can build a `SafeCdev` object without `.0` initialized properly and call it?"

From a quick look, it seems the only way to build a `SafeCdev` object is through `alloc()`, which ensures it is a valid pointer, and it never gets modified. So you could write something like this.

However, this is one of those cases where what we have is a type invariant. For these, we started the convention of writing a `# Type invariants` section in the type doc-comment, and then referencing that in the `SAFETY` comments. Doing it this way makes things much easier to reason about, and we avoid having to repeat the same explanation in several places (e.g. every time we dereference the pointer member).

Examples of type invariants being used are at `pages.rs`, `file_operations.rs` and `sync/arc.rs`:

```rust
/// Wraps the kernel's `struct file`.
///
/// # Invariants
///
/// The pointer [`File::ptr`] is non-null and valid.
pub struct File {
    ptr: *const bindings::file,
}
```

I think I will put this explanation somewhere in `coding.rst`, since it will probably come up several times :)

Reply:
Sorry -- I was writing this second comment after submitting the other; didn't mean to leave you hanging :P

Reply:
Hey, don't worry about stuff like that! And thank you so much for the great explanation, this should give a chance of getting it right.

Reply:
You're welcome!

Also, please don't forget to add a `// INVARIANTS: ...` comment in the place you actually build (or modify) the object, explaining why the invariant starts to hold (or still holds, respectively). See the files I mentioned for examples or that.